src/psyclone/psyir/nodes/operation.py

```diff
@@ -224,38 +242,209 @@ def create(operator, lhs, rhs):
-        if not isinstance(operator, Enum) or \
-           operator not in BinaryOperation.Operator:
+        if (not isinstance(operator, Enum) or
+                operator not in BinaryOperation.Operator):
             raise GenerationError(
                 f"operator argument in create method of BinaryOperation class "
                 f"should be a PSyIR BinaryOperation Operator but found "
                 f"'{type(operator).__name__}'.")
-
         binary_op = BinaryOperation(operator)
         binary_op.addchild(lhs)
         binary_op.addchild(rhs)
         return binary_op
 
+    def _get_result_precision(self, precisions):
+        '''
+        Compares the two precisions to determine the precision of the result
+        of the operation.
+
+        If the two precisions are the same, then that value is returned.
+        Otherwise, Section 7.1.9.3 of the Fortran standard says that in this
+        case, the precision of the result is the greater of the two.
+        If the precision cannot be determined then
+        `ScalarType.Precision.UNDEFINED` is returned.
+
+        :param precisions: the precision of the two operands.
+        :type precisions: list[int |
+            :py:class:`psyclone.psyir.symbols.ScalarType.Precision |
+            :py:class:`psyclone.psyir.nodes.Reference`]
+
+        :returns: the precision of the result of the operation.
+        :rtype: int | :py:class:`psyclone.psyir.symbols.ScalarType.Precision
+
+        :raises InternalError: if an unsupported Precision value is encountered
+            (this is to defend against any future extension of
+            ScalarType.Precision).
+
+        '''
+        if precisions[0] == precisions[1]:
+            return precisions[0]
+
+        # Operands have different precisions.
+        if all(isinstance(prec, int) for prec in precisions):
+            # Both precisions are integer.
+            return max(precisions)
+
+        if all(isinstance(prec, ScalarType.Precision) for
+               prec in precisions):
+            # Both precisions are of ScalarType.Precision type.
+            if ScalarType.Precision.UNDEFINED in precisions:
+                return ScalarType.Precision.UNDEFINED
+            if ScalarType.Precision.DOUBLE in precisions:
+                return ScalarType.Precision.DOUBLE
+            raise InternalError(
+                f"Operation._get_result_precision: got unsupported Precision "
+                f"value(s) '{precisions[0]}' and '{precisions[1]}' for "
+                f"operands '{self.children[0].debug_string()}' and "
+                f"'{self.children[1].debug_string()}'")
+
+        # We can't reason about the precision of the result.
+        return ScalarType.Precision.UNDEFINED
+
+    def _get_result_scalar_type(self, argtypes):
+        '''
+        Examines the two operand types to determine the base type of the
+        operation. If the type cannot be determined then an instance of
+        `DeferredType` is returned.
```

Comment:
Maybe mention that this follows "Section 7.2 of the Fortran2008" as in the equivalent precision method docstring.

Reply:
Done.